vimode/src/utils.c

```diff
@@ -219,3 +220,34 @@ void goto_nonempty(ScintillaObject *sci, gint line, gboolean scroll)
 		pos = NEXT(sci, pos);
 	SET_POS(sci, pos, scroll);
 }
+
+
+void ensure_current_line_expanded(ScintillaObject *sci)
+{
+	gint line = GET_CUR_LINE(sci);
+	if (!SSM(sci, SCI_GETLINEVISIBLE, line, 0))
+		SSM(sci, SCI_ENSUREVISIBLE, line, 0);
+}
+
+
+gint jump_to_expended_parent(ScintillaObject *sci, gint line)
+{
+	gint fold_parent = line;
+
+	/* go through the parents as long as they are not visible */
+	while (SSM(sci, SCI_GETLINEVISIBLE, fold_parent, 0) == FALSE) {
+		gint prev_parent = SSM(sci, SCI_GETFOLDPARENT, fold_parent, 0);
+
+		if (prev_parent == -1) break;
+		fold_parent = prev_parent;
+	}
+
+	if (fold_parent != line)
+	{
+		/* move the cursor on the visible line before the fold */
+		gint pos = SSM(sci, SCI_POSITIONFROMLINE, fold_parent, 0);
+		SET_POS(sci, pos, TRUE);
```

Comment:
Use `SET_POS_NOX()` instead. This one will preserve the "maximum x" cursor value so when further moving up or down, this one will be recovered.